improver/utilities/threshold_interpolation.py

```diff
@@ -0,0 +1,199 @@
+        forecast_at_thresholds, coord=threshold_coord_name
+    )
+
+    forecast_at_interpolated_thresholds = interpolate_multiple_rows_same_x(
+        np.array(desired_thresholds, dtype=np.float64),
+        original_thresholds.astype(np.float64),
+        forecast_at_reshaped_thresholds.astype(np.float64),
+    )
+
+    forecast_at_interpolated_thresholds = np.transpose(
+        forecast_at_interpolated_thresholds
+    )
+
+    # Reshape forecast_at_percentiles, so the percentiles dimension is
+    # first, and any other dimension coordinates follow.
+    forecast_at_thresholds_data = restore_non_percentile_dimensions(
+        forecast_at_interpolated_thresholds,
+        next(forecast_at_thresholds.slices_over(threshold_coord_name)),
+        len(desired_thresholds),
+    )
+
+    template_cube = next(forecast_at_thresholds.slices_over(threshold_coord_name))
+    template_cube.remove_coord(threshold_coord_name)
+    threshold_cube = create_cube_with_thresholds(
+        desired_thresholds,
+        template_cube,
+        forecast_at_thresholds_data,
+        threshold_coord_name,
+    )
+
+    if original_mask is not None:
+        original_mask = np.broadcast_to(original_mask, threshold_cube.shape)
+        threshold_cube.data = np.ma.MaskedArray(threshold_cube.data, mask=original_mask)
+
+    return threshold_cube
+
+
+def create_cube_with_thresholds(
+    thresholds: Union[List[float], ndarray],
+    template_cube: Cube,
+    cube_data: ndarray,
+    threshold_coord_name: str,
+) -> Cube:
+    """
+    Create a cube with a threshold coordinate based on a template cube.
+    The resulting cube will have an extra threshold coordinate compared with
+    the template cube. The shape of the cube_data should be the shape of the
+    desired output cube.
+
+    Args:
+        thresholds:
+            Ensemble thresholds. There should be the same number of
+            thresholds as the first dimension of cube_data.
+        template_cube:
+            Cube to copy metadata from.
+        cube_data:
+            Data to insert into the template cube.
+            The shape of the cube_data, excluding the dimension associated with
+            the threshold coordinate, should be the same as the shape of
+            template_cube.
+            For example, template_cube shape is (3, 3, 3), whilst the cube_data
+            is (10, 3, 3, 3), where there are 10 thresholds.
+        cube_unit:
+            The units of the data within the cube, if different from those of
+            the template_cube.
+
+    Returns:
+        Cube containing a percentile coordinate as the leading dimension (or
+        scalar percentile coordinate if single-valued)
+    """
+    # create cube with new threshold dimension
+    cubes = iris.cube.CubeList([])
+    for point in thresholds:
+        cube = template_cube.copy()
+        coord = iris.coords.DimCoord(
+            np.array([point], dtype="float32"), units=cube.units
+        )
```

Comment:
This threshold coordinate also needs a relative to attribute to specify that the probabilities are for values less than the thresholds.